collections/index.js

```diff
@@ -20,46 +20,130 @@ import employees from './data.json';
+    // in each of the iterations, if the employee 'department' matches the
+    // provided arguement 'dept', the statement is true...
+    if (employees[employeeNumberIndex].department === dept) {
+      // this adds (pushes) the employees details to the departmentEmployees
+      // array we created earlier.
+      departmentEmployees.push(employees[employeeNumberIndex]);
+    }
+  }
+  // returns all employees that match the department or an empty array if
+  // there were no matches.
+  return departmentEmployees;
+  // side note: i wrote this completely before testing and it worked first time!
 }
 
 /*
   I expect this function to look through all the employees and return an array
   of employees who are missing any data
 */
-export const getEmployeesWithMissingData = () => {
 
+export const getEmployeesWithMissingData1 = () => {
+  // sets an empty array ready to accept any values we add later:
+  var missingInfoEmployees = [];
+  // using a for loop to index through each of the employees on record
+  for (let employeeNumberIndex = 0; employeeNumberIndex < employees.length; employeeNumberIndex++) {
+    // checks if any of the values are empty. I prefer this method over others
+    // I thought of because if more than one value is empty per employee, this
+    // function only adds the employee to the array once.
+    if (employees[employeeNumberIndex].id === "" || employees[employeeNumberIndex].firstName === "" || employees[employeeNumberIndex].lastName === "" || employees[employeeNumberIndex].department === "" || employees[employeeNumberIndex].contactNumber === "" || employees[employeeNumberIndex].taxCode === "" || employees[employeeNumberIndex].salary === "") {
+      // this adds (pushes) the employees details to the missingInfoEmployees
+      // array we created earlier.
+      missingInfoEmployees.push(employees[employeeNumberIndex]);
+    }
+  }
+  // returns all employees that are missing data
+  return missingInfoEmployees;
+}
+
+// I thought of a 2nd way of doing this which is probably better but might
+// need some tidying up.
+
+export const getEmployeesWithMissingData = () => {
+  // sets an empty array ready to accept any values we add later:
+  var missingInfoEmployees = [];
+  // using a for loop to index through each of the employees on record
+  for (let employeeNumberIndex = 0; employeeNumberIndex < employees.length; employeeNumberIndex++) {
+    // creates an array of the employee details
+    let employeeDetails = Object.values(employees[employeeNumberIndex])
+    // checks each value in the new array for any empty strings
+    employeeDetails.find(function(item) {
+      if (item === "") {
+        // adds the employee to the missingInfoEmployees array
+        missingInfoEmployees.push(employees[employeeNumberIndex]);
+      }
+    })
+  }
+  // returns all employees that are missing data
+  return missingInfoEmployees;
```

Comment:
Really good shit - See how much cleaner using `Object.values()` and then `array.find()` is rather than checking for each individual property. The issue with doing the first way is that what if the user object has 1000 properties? you have to write 1000 conditionals in the if statement - however this solution will work for any number of properties. 🔥